### ISSUE TITLE
RDF Classes

### ISSUE THREAD

**Issue by davaya** [spdx/spdx-3-model#423]
As a follow-up to the July 14 serialization discussion of #368, [RDFS](https://www.w3.org/TR/rdf11-schema/#ch_classes) says:

> 2.2 rdfs:Class
This is the class of resources that are RDF classes. rdfs:Class is an instance of rdfs:Class.
> 2.3 rdfs:Literal
The class rdfs:Literal is the class of [literal](http://www.w3.org/TR/rdf11-concepts/#section-Graph-Literal) values such as strings and integers. Property values such as textual strings are examples of RDF literals.
rdfs:Literal is an instance of [rdfs:Class](https://www.w3.org/TR/rdf11-schema/#ch_class). rdfs:Literal is a [subclass](https://www.w3.org/TR/rdf11-schema/#def-subclass) of [rdfs:Resource](https://www.w3.org/TR/rdf11-schema/#ch_resource).
> 2.4 rdfs:Datatype
rdfs:Datatype is the class of datatypes. All instances of rdfs:Datatype correspond to the [RDF model of a datatype](http://www.w3.org/TR/rdf11-concepts/#section-Datatypes) described in the RDF Concepts specification [[RDF11-CONCEPTS](https://www.w3.org/TR/rdf11-schema/#bib-RDF11-CONCEPTS)]. rdfs:Datatype is both an instance of and a [subclass](https://www.w3.org/TR/rdf11-schema/#def-subclass) of [rdfs:Class](https://www.w3.org/TR/rdf11-schema/#ch_class). Each instance of rdfs:Datatype is a [subclass](https://www.w3.org/TR/rdf11-schema/#def-subclass) of rdfs:Literal.

RDF Concepts says:

> 5. Datatypes
Datatypes are used with RDF [literals](https://www.w3.org/TR/rdf11-concepts/#dfn-literal) to represent values such as strings, numbers and dates. The datatype abstraction used in RDF is compatible with XML Schema [[XMLSCHEMA11-2](https://www.w3.org/TR/rdf11-concepts/#bib-XMLSCHEMA11-2)].

and XML Schema says:

> 3.3.1.3 Facets
The [string](https://www.w3.org/TR/xmlschema11-2/#string) datatype has the following [·constraining facets·](https://www.w3.org/TR/xmlschema11-2/#dt-constraining-facet) with the values shown; these facets may be specified in the derivation of new types, if the value given is at least as restrictive as the one shown:
[whiteSpace](https://www.w3.org/TR/xmlschema11-2/#rf-whiteSpace) = preserve
Datatypes derived by restriction from [string](https://www.w3.org/TR/xmlschema11-2/#string) may also specify values for the following [·constraining facets·](https://www.w3.org/TR/xmlschema11-2/#dt-constraining-facet):

[length](https://www.w3.org/TR/xmlschema11-2/#rf-length)
[minLength](https://www.w3.org/TR/xmlschema11-2/#rf-minLength)
[maxLength](https://www.w3.org/TR/xmlschema11-2/#rf-maxLength)
[pattern](https://www.w3.org/TR/xmlschema11-2/#rf-pattern)
[enumeration](https://www.w3.org/TR/xmlschema11-2/#rf-enumeration)
[assertions](https://www.w3.org/TR/xmlschema11-2/#rf-assertions)

-----

What that says to me is that xsd:string is a Datatype, which is a Class, and that new types derived from it using model-defined facets are also Classes identified by name.

The markdown files can have a Datatypes directory separate from the Classes directory to hold `pattern`-restricted classes, just as the separate Vocabularies directory holds `enumeration`-restricted classes.  For uniformity of naming the model should define datatype names for the built-in [primitive datatypes](https://www.w3.org/TR/xmlschema11-2/#built-in-primitive-datatypes) that we use.  Conceptually there is no difference:
> The [·built-in·](https://www.w3.org/TR/xmlschema11-2/#dt-built-in) [·constructed·](https://www.w3.org/TR/xmlschema11-2/#dt-constructed) datatypes are those which are believed to be so common that if they were not defined in this specification many schema designers would end up reinventing them.

So if a property can have type DateTime, it might as well also be able to have type String instead of xsd:string.

**Comment by goneall** [comment]
Although technically Datatype is a class, most tooling I've used in defining and maintaining RDF ontologies clearly distinguish general "object classes" from Datatypes - so having a separate Datatype directory would be consistent with some of the common practices.

In terms of the exact format, I did some experiments using Protege to define a new Datatype.

What I came up with was defining a Datatype with a comment annotation containing the pattern restriction and a definition using the built-in xsd:string.  This results in the following Datatype definition in RDF/OWL/TTL format:

```
:test rdf:type rdfs:Datatype ;
      rdfs:comment "Restricted by the pattern \".*\"" ;
      owl:equivalentClass xsd:string .
```

**Comment by goneall** [comment]
> So if a property can have type DateTime, it might as well also be able to have type String instead of xsd:string.

I'm not sure I understand the benefit of the above statement.

I believe we should use the built-in XSD datatype wherever possible - e.g. continue to use xsd:string.

**Comment by davaya** [comment]
The type name is a synonym for the built-in xsd type, so one can be translated to the other where it makes sense as long as the translation `xsd:string = String` is defined somewhere..  It could be hidden in software or made explicit in the model files, I just prefer explicit.

The benefit is a consistent language-independent representation of the model - xsd is specific to XML while other formats have their own representations of types.  If CreationInfo and Relationship and DateTime are all types in a model regardless of what languages it targets, it's just aesthetically more consistent for SemVer and String to have names in the same style.

**Comment by zvr** [comment]
I am against the idea that we create our own "language-independent representation of the model".

We have decided long ago that we use RDF for this model. The only point of the markdown is to make it accessible to people who do not want to deal with RDF intricacies.

**Comment by davaya** [comment]
@goneall since you are using protege, [Stackoverflow](https://stackoverflow.com/questions/24531940/defining-datarange-expression-in-protege-for-a-data-type-property) had the following datatype definition (for number with range 0-100) - presumably the same Datatype owl:withRestrictions structure works for strings with a pattern ...

```
<rdf:RDF
    xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#"
    xmlns:owl="http://www.w3.org/2002/07/owl#"
    xmlns="http://stackoverflow.com/q/24531940/1281433/percentages#"
    xmlns:rdfs="http://www.w3.org/2000/01/rdf-schema#"
    xmlns:xsd="http://www.w3.org/2001/XMLSchema#">
  <owl:Ontology rdf:about="http://stackoverflow.com/q/24531940/1281433/percentages"/>
  <owl:DatatypeProperty rdf:about="http://stackoverflow.com/q/24531940/1281433/percentages#hasPercentage">
    <rdfs:range>
      <rdfs:Datatype>
        <owl:onDatatype rdf:resource="http://www.w3.org/2001/XMLSchema#double"/>
        <owl:withRestrictions rdf:parseType="Collection">
          <rdf:Description>
            <xsd:minInclusive rdf:datatype="http://www.w3.org/2001/XMLSchema#integer"
            >0</xsd:minInclusive>
          </rdf:Description>
          <rdf:Description>
            <xsd:maxInclusive rdf:datatype="http://www.w3.org/2001/XMLSchema#integer"
            >100</xsd:maxInclusive>
          </rdf:Description>
        </owl:withRestrictions>
      </rdfs:Datatype>
    </rdfs:range>
  </owl:DatatypeProperty>
</rdf:RDF>
```

**Comment by zvr** [comment]
@davaya I do not understand your point at all. What does `xsd:string` have to do with Vocabularies?

In case it's not yet clear, the value of the `relationshipType` type property is an object of class `RelationshipType` (or `http://spdx.org/rdf/v3/Core/RelationshipType` to give its full URI).

One such individual is `http://spdx.org/rdf/v3/Core/RelationshipType/contains`.
There are no strings anywhere.

There might be a serialization that uses a string like `"CONTAINS"`, as the tag-value serialization did in SPDXv2, but this has nothing to do with the model.

**Comment by davaya** [comment]
[Enumerations](https://www.w3.org/TR/xmlschema11-2/#rf-enumeration) are a restriction on the xsd:string datatype.  [Patterns](https://www.w3.org/TR/xmlschema11-2/#rf-pattern) are a restriction on the xsd:string datatype.  The lexical space of an instance of RelationshipType is a string such as "affects", "amends", "ancestor", etc.  Only certain strings are valid instances of the RelationshipType class, just as only certain strings are valid instances of the DateTime class defined by a pattern restriction on xsd:string.

As I said at the top, the [datatype](https://www.w3.org/TR/rdf11-concepts/#section-Datatypes) abstraction used in RDF is compatible with XML Schema

**Comment by goneall** [comment]
@davaya Is this now resolved with the introduction of the DataType directory?

**Comment by davaya** [comment]
The existence of Datatypes and Vocabularies directories are fine, if the problem was the inability of software to properly process the markdown files without those directories.  (Strictly speaking, following this precedent, there would need to be both SimpleDatatypes (for DateTime, MediaType, etc classes) and Datatypes (for CreationInfo, PositiveIntegerRange, etc classes), but I'm not advocating doing that.)

Classes, simple Datatypes, compound Datatypes, and Vocabularies are all Classes, but whatever markdown directories they are placed in is OK.  Model parsers should check that compound Datatypes do not have a SubclassOf chain with Element as the root (i.e. that compound Datatypes are not Elements and aren't identified by SpdxId), but since they are currently in the Classes directory, their non-Element nature is what distinguishes Element classes from compound Datatype classes.

My goal is to stop the nitpicking over terminology.  "SubclassOf" means "inherit from" whether it is a simple datatype restricted from another simple datatype, an Element inherited from another Element, or a compound Datatype inherited from another compound Datatype.

**TL;DR:** it's resolved, with or without the Datatypes directory, if we stop arguing about the meaning of `SubclassOf`.

**Comment by davaya** [comment]
The existence of Datatypes and Vocabularies directories are fine, if the problem was the inability of software to properly process the markdown files without those directories.  (Strictly speaking, following this precedent, there would need to be both SimpleDatatypes (for DateTime, MediaType, etc classes) and Datatypes (for CreationInfo, PositiveIntegerRange, etc classes), but I'm not advocating doing that.)

Classes, simple Datatypes, compound Datatypes, and Vocabularies are all Classes, but whatever markdown directories they are placed in is OK.  Model parsers should check that compound Datatypes do not have a SubclassOf chain with Element as the root (i.e. that compound Datatypes are not Elements and aren't identified by SpdxId), but since they are currently in the Classes directory, their non-Element subclass tree is what distinguishes Element classes from compound Datatype classes.

My goal is to stop the nitpicking over terminology.  "SubclassOf" means "inherit from" whether it is a simple datatype restricted from another simple datatype, an Element inherited from another Element, or a compound Datatype inherited from another compound Datatype.

**TL;DR:** it's resolved, with or without the Datatypes directory, if we stop arguing about the meaning of `SubclassOf`.
It's resolved now that we have accepted that DateTime, MediaType, and SemVer can have names without arguing about how many times they are used.
And it's resolved since we still have PR #407 to answer the question raised in Issue #36.  That question is as simple to resolve as the DateTime question - does giving SpdxId a name make the model easier to understand, and does it cause any harm.  Arguing about whether giving names to DateTime or SpdxId is "necessary" is a waste of time - they aren't "necessary", but the names are helpful and not harmful.

**Comment by goneall** [comment]
@davaya If I understand you comment above, we should be able to close this issue and leave PR #407 open since 407 is more specific and actionable.  We can always refer back to this issue and/or create another PR with a specific proposal.
